File: Probs_including_Gichul/PPFDS/day11_1122_heap_and_pq/lec8/selfstudy/heap_array.py
class PriorityQueueArray:
    def __init__(self, is_min=True):
        self.heap = []
        self.is_min = is_min  # True for min-heap, False for max-heap

    def parent(self, i):
        return (i - 1) // 2

    def left_child(self, i):
        return 2 * i + 1

    def right_child(self, i):
        return 2 * i + 2

    def enqueue(self, key):
        # TODO: Add element and maintain heap property
        self.heap.append(key)
        n = len(self.heap)-1
        #bubble up
        while n != 0 and self.heap[self.parent(n)] > self.heap[n]:
            self.heap[self.parent(n)],self.heap[n] = self.heap[n],self.heap[self.parent(n)]
            n = self.parent(n)

    def dequeue(self):
        # TODO: Remove and return top element while maintaining heap property
        temp = self.heap[0]
        n = len(self.heap)
        self.heap[0],self.heap[n-1] = self.heap[n-1],self.heap[0]
        self.heap.pop()
        self.heapify(0)
        return temp
    
    def heapify(self, i):
        # TODO: Maintain heap property at index i
        # min-heap
        if self.is_min:
            smallest = i
            left = self.left_child(i)
            right = self.right_child(i)

            if left<len(self.heap) and self.heap[i] > self.heap[left]:
                smallest = left
            if right<len(self.heap) and self.heap[smallest] > self.heap[right]:
                smallest = right
            if smallest != i:
                self.heap[i],self.heap[smallest] = self.heap[smallest],self.heap[i]
                self.heapify(smallest)
        # max는 else 후 반대로

    def print_queue(self):
        print(self.heap)

# Test case
pq = PriorityQueueArray(is_min=True)
pq.enqueue(10)
pq.enqueue(5)
pq.enqueue(20)
pq.enqueue(1)
pq.print_queue()
print(pq.dequeue())
pq.print_queue()

File: Probs_including_Gichul/PPFDS/day11_1122_heap_and_pq/lec8/selfstudy/heap_array.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_SIZE 100

typedef struct {
    int data[MAX_SIZE]; // Array to store heap elements
    int size;           // Current number of elements in the heap
} MaxHeapArray;

// Function prototypes
void swap(int* a, int* b){
    int temp = *a;
    *a = *b;
    *b = temp;
}

void init_heap(MaxHeapArray* heap);
void insert(MaxHeapArray* heap, int key);
int extract_max(MaxHeapArray* heap);
void heapify(MaxHeapArray* heap, int index);
void print_heap(MaxHeapArray* heap);
// Function definitions (to be implemented)
void init_heap(MaxHeapArray* heap) {
    // TODO: Initialize the heap structure
    heap->size = 0;
}

void insert(MaxHeapArray* heap, int key) {
    // TODO: Add an element to the heap and restore max-heap property
    heap->data[heap->size] = key;
    heap->size++;
    for(int i=heap->size-1;i>=0;i--){
        heapify(heap,i);
    }
}

int extract_max(MaxHeapArray* heap) {
    // TODO: Remove and return the maximum element (root)
    if(heap->size == 0) return -1;
    else{
        int ans = heap->data[0];
        heap->data[0] = heap->data[heap->size-1];
        // heap->data[heap->size-1] = NULL;
        heap->size--;
        heapify(heap,0);
        return ans;
    }
}

void heapify(MaxHeapArray* heap, int index) {
    // TODO: Restore max-heap property from the given index
    int largest = index;
    int left = index*2+1;
    int right = index*2+2;
    int n = heap->size;

    if(left < n && heap->data[left] > heap->data[index]) largest = left;
    if(right < n && heap->data[right] > heap->data[largest]) largest = right;

    if(largest != index){
        swap(&(heap->data[index]),&(heap->data[largest]));
        heapify(heap,largest);
    }
}

void print_heap(MaxHeapArray* heap) {
    // TODO: Print the elements of the heap
    for(int i=0;i<heap->size;i++){
        printf("%d ",heap->data[i]);
    }
    printf("\n");
}

// Main function to test the logic
int main() {
    MaxHeapArray heap;
    init_heap(&heap);

    // Test cases
    insert(&heap, 10);
    insert(&heap, 5);
    insert(&heap, 20);
    insert(&heap, 1);
    insert(&heap, 7);
    insert(&heap, 11);
    insert(&heap, 30);

    printf("Heap after insertions: ");
    print_heap(&heap);

    printf("Extracted Max: %d\n", extract_max(&heap));
    printf("Heap after extraction: ");
    print_heap(&heap);

    return 0;
}